mlir-tensorrt/compiler/include/mlir-tensorrt/Compiler/OptionsProviders.h

```diff
@@ -0,0 +1,121 @@
+#ifndef MLIR_TENSORRT_COMPILER_OPTIONS
+#define MLIR_TENSORRT_COMPILER_OPTIONS
+
+#include "mlir-tensorrt-dialect/Utils/Options.h"
+#include "mlir/Support/LLVM.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Error.h"
+#include <string>
+
+namespace mlirtrt::compiler {
+
+struct OptionsProvider {
+  virtual ~OptionsProvider() {}
+  virtual void addToOptions(mlir::OptionsContext &context) = 0;
+
+  virtual llvm::Error finalize() { return llvm::Error::success(); }
+};
+
+/// DebugOptions are options that are common to different compiler API
+/// interfaces.
+struct DebugOptions : public OptionsProvider {
+  /// A directory path where the IR will be dumped during compilation
+  /// using the `mlir-print-ir-tree-dir` mechanism.
+  std::string dumpIRPath = "";
+
+  /// Whether the LLVM 'debug' flag that enables execution of code guarded by
+  /// the `LLVM_DEBUG` macro should be set to 'on'. This results in very verbose
+  /// output from the compiler dumped to stderr.
+  bool enableLLVMDebugFlag = false;
+
+  /// A set of names to be given to the LLVM 'debug types' option, akin to
+  /// setting
+  /// `-debug-types=...` from the command line.
+  mlir::SmallVector<std::string> llvmDebugTypes = {};
+
+  void addToOptions(mlir::OptionsContext &context) override {
+    context.addOption("mlir-print-ir-tree-dir", dumpIRPath, llvm::cl::init(""));
+    context.addOption("debug", enableLLVMDebugFlag);
+    context.addList<std::string>("debug-only", llvmDebugTypes,
+                                 llvm::cl::ZeroOrMore,
+                                 llvm::cl::CommaSeparated);
+  }
+};
+
+struct ExecutorOptions : public OptionsProvider {
+  /// The host index bit-width.
+  int64_t executorIndexBitwidth{64};
+
+  /// Whether to pass memref's as struct/table in function calls.
+  bool executorUsePackedMemRefCConv{true};
+
+  void addToOptions(mlir::OptionsContext &context) override {
+    context.addOption("executor-index-bitwidth", executorIndexBitwidth,
+                      llvm::cl::init(64));
+  }
+};
+
+struct DeviceOptions : public OptionsProvider {
+  /// Target device compute capability (SM version)
+  int64_t computeCapability;
+
+  /// Target device max shared memory per block (kilobytes)
+  int64_t maxSharedMemoryPerBlockKb;
+
+  /// Target device maximum 4-byte register sper block.
+  uint64_t maxRegistersPerBlock;
+
+  void addToOptions(mlir::OptionsContext &context) override {
+    context.addOption(
+        "device-compute-capability", computeCapability, llvm::cl::init(64),
+        llvm::cl::desc("Sets the device compute capbility. Only relevant "
+                       "if '--device-infer-from-host=false'"));
+    context.addOption("device-max-shared-memory-per-block-kb",
+                      maxSharedMemoryPerBlockKb, llvm::cl::init(0));
+    context.addOption("device-max-registers-per-block", maxRegistersPerBlock,
+                      llvm::cl::init(0));
```

Comment:
Set default value to 65536 (limit for all arch since sm60).

Reply:
Updated.